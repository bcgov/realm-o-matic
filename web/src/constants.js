//
// Realm-o-Matic
//
// Copyright Â© 2019 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Shelly Xue Han
//

import config from './config.json';

export const SSO_CONFIG = {
  baseURL: process.env.REACT_APP_SSO_BASE_URL || config.ssoBaseUrl,
  realmName: process.env.REACT_APP_SSO_REALM_NAME || config.ssoRealmName,
  clientId: process.env.REACT_APP_SSO_CLIENT_ID || config.ssoClientId,
};

export const API = {
  BASE_URL: () =>
    process.env.NODE_ENV === 'development'
      ? 'http://localhost:8000/api/v1/'
      : `${window.location.origin}/api/v1/`,
  TIME_OUT: 40000,
  IDP: () => 'idps',
};

export const TEST_IDS = {
  APP: {
    LOGIG: 'login-button',
    GET_IDPS: 'get-idps-button',
  },
};

export const APP_INFO = {
  NAME: 'Realm-o-Matic',
};

export const SSO_IDP = {
  GITHUB: 'github',
  IDIR: 'idir',
};

export const FOOTER_LINKS = [
  {
    name: 'Home',
    link: '.',
  },
  {
    name: 'Disclaimer',
    link: 'https://www2.gov.bc.ca/gov/content/home/disclaimer',
  },
  {
    name: 'Privacy',
    link: 'https://www2.gov.bc.ca/gov/content/home/privacy',
  },
  {
    name: 'Accessibility',
    link: 'https://www2.gov.bc.ca/gov/content/home/accessibility',
  },
  {
    name: 'Copyright',
    link: 'https://www2.gov.bc.ca/gov/content/home/copyright',
  },
  {
    name: 'Contact Us',
    link: 'https://github.com/bcgov/realm-o-matic',
  },
];
