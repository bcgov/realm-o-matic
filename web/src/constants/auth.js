//
// Realm-o-Matic
//
// Copyright Â© 2019 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Shelly Xue Han
//

import config from '../config.json';

export const SSO_CONFIG = {
  baseURL: process.env.REACT_APP_SSO_BASE_URL || config.ssoBaseUrl,
  realmName: process.env.REACT_APP_SSO_REALM_NAME || config.ssoRealmName,
  clientId: process.env.REACT_APP_SSO_CLIENT_ID || config.ssoClientId,
};

export const SSO_IDP = {
  GITHUB: 'github',
  IDIR: 'idir',
};

export const ACCESS_CONTROL = {
  REQUESTER_ROLE: 'devhub_kc_requester',
  REVIEWER_ROLE: 'devhub_kc_reviewer',
  NO_ROLE: 'unauthorized',
};
