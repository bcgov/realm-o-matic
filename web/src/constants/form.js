//
// Realm-o-Matic
//
// Copyright Â© 2019 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Shelly Xue Han
//

const basicInfo = {
  name: 'realmInfo',
  title: 'Realm Information',
  elements: [
    {
      type: 'text',
      name: 'realmId',
      title: 'The Unique ID of your realm',
      isRequired: true,
      enableIf: 'false',
    },
    {
      type: 'text',
      name: 'displayName',
      title: 'The realm display name',
      isRequired: true,
    },
    {
      type: 'text',
      name: 'adminUser',
      title: 'The administrator of the realm (IDIR username)',
      isRequired: true,
    },
    {
      type: 'checkbox',
      name: 'idps',
      title: 'Required identity providers',
      isRequired: true,
      choices: ['IDIR', 'GitHub', 'BCeID'],
    },
    {
      type: 'text',
      name: 'po',
      title: 'The product owner of project',
      isRequired: true,
      validators: [
        {
          type: 'email',
        },
      ],
    },
  ],
};

const requesterInfo = {
  name: 'requesterInfo',
  title: 'Requester Information',
  elements: [
    {
      type: 'panel',
      name: 'requester',
      elements: [
        {
          type: 'text',
          name: 'requesterEmail',
          title: 'Email',
          isRequired: true,
          enableIf: 'false',
          colCount: 0,
          validators: [
            {
              type: 'email',
            },
          ],
        },
        {
          type: 'text',
          name: 'requesterIDIR',
          title: 'IDIR Username',
          // startWithNewLine: false,
          isRequired: true,
          enableIf: 'false',
          colCount: 0,
        },
        {
          type: 'text',
          name: 'requesterLastName',
          title: 'Last Name (optional)',
          colCount: 0,
        },
        {
          type: 'text',
          name: 'requesterFirstName',
          title: 'First Name (optional)',
          colCount: 0,
        },
      ],
      title: 'Requester Information',
    },
  ],
};

export const formJson = {
  showProgressBar: 'top',
  showQuestionNumbers: 'off',
  pages: [basicInfo, requesterInfo],
};

export const formDataToRequest = {
  id: 'id',
  realm: {
    id: 'realmId',
    displayName: 'displayName',
    adminUser: 'adminUser',
    idps: 'idps',
    po: 'po',
  },
  requester: {
    username: 'requesterIDIR',
    email: 'requesterEmail',
    firstName: 'requesterFirstName',
    lastName: 'requesterLastName',
  },
};

export const requestToFormData = {
  id: 'id',
  realmId: 'realm.id',
  displayName: 'realm.displayName',
  requesterEmail: 'requester.email',
  requesterIDIR: 'requester.username',
  requesterLastName: 'requester.lastName',
  requesterFirstName: 'requester.firstName',
  adminUser: 'realm.adminUser',
  idps: 'realm.adminUser',
  po: 'realm.po',
};
