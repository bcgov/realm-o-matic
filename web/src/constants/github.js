//
// Realm-o-Matic
//
// Copyright Â© 2019 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Shelly Xue Han
//

export const REQUEST_STATUS = {
  PROCESSING: {
    text: 'Processing',
    color: 'olive',
  },
  SUCCESS: {
    text: 'Success',
    color: 'green',
  },
  FAILED: {
    text: 'Failed',
    color: 'red',
  },
  // for BCeID flow:
  PENDING: {
    text: 'Pending',
    color: 'orange',
  },
  REJECT: {
    text: 'Rejected',
    color: 'red',
  },
  UNKNOWN: {
    text: 'unknown',
    color: 'grey',
  },
};

export const GITHUB_LABELS = {
  READY: 'request-ready',
  FAILED: 'request-failed',
  COMPLETED: 'realm-created',
  BCEID: 'bceid-requested',
  BCEID_APPROVED: 'bceid-approved',
  BCEID_REJECTED: 'bceid-rejected',
  BCEID_COMPLETED: 'bceid-enabled',
};

export const GITHUB_PR_STATUS = {
  OPEN: 'open',
  CLOSED: 'closed',
};

export const getPrStatus = (state, merged, labels = []) => {
  const isMerged = merged ? true : false;
  switch (state) {
    case GITHUB_PR_STATUS.OPEN:
      // failed:
      if (labels.includes(GITHUB_LABELS.FAILED)) return REQUEST_STATUS.FAILED;
      // bceid rejected:
      else if (labels.includes(GITHUB_LABELS.BCEID_REJECTED)) return REQUEST_STATUS.REJECT;
      // when BCeID is requested, but not yet approved or completed, request is pending:
      else if (labels.includes(GITHUB_LABELS.BCEID) && !labels.includes(GITHUB_LABELS.BCEID_APPROVED) && !labels.includes(GITHUB_LABELS.BCEID_COMPLETED)) return REQUEST_STATUS.PENDING;
      // otherwise, before the PR is closed, some tasks is processing:
      else return REQUEST_STATUS.PROCESSING;
    case GITHUB_PR_STATUS.CLOSED:
      return isMerged ? REQUEST_STATUS.SUCCESS : REQUEST_STATUS.FAILED;
    default:
      // throw Error(`Unknown PR state: ${state}!`);
      return REQUEST_STATUS.UNKNOWN;
  }
};
