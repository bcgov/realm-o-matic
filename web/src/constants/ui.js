//
// Realm-o-Matic
//
// Copyright Â© 2019 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Shelly Xue Han
//

export const TEST_IDS = {
  AUTH: {
    MODAL: 'login-modal',
    IDIR_LOGIN: 'idir-login-button',
    GITHUB_LOGIN: 'github-login-button',
  },
  APP: {
    NAME: 'app-name',
    LOGO: 'gov-logo',
    ACCOUNT: 'account-dropdown',
    NEW_REQUEST: 'new-request-button',
    HIDDEN_NEW_REQUEST: 'hidden-new-request-button',
    LOADER: 'load-indicator',
    EMPTY: 'empty-content',
    POPUP: 'status-popup',
  },
  REQUEST: {
    MESSAGE: 'request-message',
    FORM: 'request-form',
    FORM_LIST: 'request-list',
    APPROVAL: 'request-approval',
    REJECT: 'request-rejection',
    CANCEL: 'go-back',
  },
};

export const APP_INFO = {
  NAME: 'Realm-o-Matic',
  CONTACT_LINK: 'https://chat.pathfinder.gov.bc.ca',
};

export const FOOTER_LINKS = [
  {
    name: 'Home',
    link: '.',
  },
  {
    name: 'Disclaimer',
    link: 'https://www2.gov.bc.ca/gov/content/home/disclaimer',
  },
  {
    name: 'Privacy',
    link: 'https://www2.gov.bc.ca/gov/content/home/privacy',
  },
  {
    name: 'Accessibility',
    link: 'https://www2.gov.bc.ca/gov/content/home/accessibility',
  },
  {
    name: 'Copyright',
    link: 'https://www2.gov.bc.ca/gov/content/home/copyright',
  },
  {
    name: 'Contact Us',
    link: 'https://github.com/bcgov/realm-o-matic',
  },
];
