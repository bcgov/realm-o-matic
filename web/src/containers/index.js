export { default as Home } from './Home';
export { default as Request } from './Request';
export { Restricted } from './Restricted';
