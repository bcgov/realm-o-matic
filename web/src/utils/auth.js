import { ImplicitAuthManager } from '@bcgov/common-web-utils';
import { SSO_CONFIG } from '../constants/auth';

const iam = new ImplicitAuthManager(SSO_CONFIG);

export default iam;
