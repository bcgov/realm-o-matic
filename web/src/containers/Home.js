import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Button, Loader } from 'semantic-ui-react';
import styled from '@emotion/styled';
import { TEST_IDS } from '../constants/ui';
import { RequestList } from '../components/Request/RequestList';
import { getRequestsAction } from '../actionCreators';

const StyledLoader = styled(Loader)`
  font-size: 1rem;
  padding: 0;
`;

export class Home extends Component {
  static displayName = '[Component Home]';

  componentDidMount = () => {
    this.props.getRequestsAction({ user: this.props.userId });
  };

  render() {
    const { requests, getRequestsStarted, errorMessage } = this.props;
    let requestsList = null;
    if (requests === null || getRequestsStarted) {
      requestsList = (
        <StyledLoader
          active
          size="small"
          inline="centered"
          content="Loading requests..."
          data-testid={TEST_IDS.APP.LOADER}
        />
      );
    } else if (requests.length === 0) {
      requestsList = (
        <p data-testid={TEST_IDS.APP.EMPTY}>
          You do not have any realm requests yet, start a new request.
        </p>
      );
    } else {
      requestsList = <RequestList requests={requests} />;
    }

    return (
      <div>
        <h2>Welcome!</h2>
        <Link to="/Request/new" data-testid={TEST_IDS.APP.NEW_REQUEST}>
          <Button data-testid="startRequest">Start Request</Button>
        </Link>
        <p>{errorMessage}</p>
        {requestsList}
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    isAuthenticated: state.authentication.isAuthenticated,
    email: state.authentication.email,
    userId: state.authentication.userId,
    // get Requests:
    errorMessage: state.getRequests.errorMessage,
    requests: state.getRequests.requests,
    getRequestsStarted: state.getRequests.getRequestsStarted,
  };
};

const mapDispatchToProps = dispatch => {
  return bindActionCreators(
    {
      getRequestsAction,
    },
    dispatch
  );
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Home);
