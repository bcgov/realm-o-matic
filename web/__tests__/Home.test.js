import React from 'react';
import 'jest-dom/extend-expect';
import { render } from 'react-testing-library';
import { BrowserRouter as Router } from 'react-router-dom';
import { TEST_IDS } from '../src/constants/ui';
import { Home } from '../src/containers/Home';
import { ACCESS_CONTROL } from '../src/constants/auth';

describe('Home Component', () => {
  const REQUESTS = [
    {
      number: '123',
      realmName: 'realm 1',
      state: 'open',
      prContent: {
        realmId: 'fewgfw123',
        requester: {
          email: '1@email.com',
        },
      },
    },
    {
      number: '345',
      realmName: 'realm 2',
      state: 'closed',
      prContent: {
        realmId: 'fewgfw345',
        requester: {
          email: '1@email.com',
        },
      },
    },
    {
      number: '777',
      realmName: 'realm 3',
      state: 'open',
      prContent: {
        realmId: 'fewgfw777',
        requester: {
          email: '1@email.com',
        },
      },
    },
  ];

  const AUTH_PROPS_0 = {
    isAuthenticated: true,
    userInfo: {
      email: '123@email.com',
    },
    userId: '123',
    errorMessage: null,
    authCode: ACCESS_CONTROL.REVIEWER_ROLE,
  };

  const AUTH_PROPS_1 = {
    isAuthenticated: true,
    userInfo: {
      email: '123@email.com',
    },
    userId: '123',
    errorMessage: null,
    authCode: ACCESS_CONTROL.REQUESTER_ROLE,
  };

  const AUTH_PROPS_3 = {
    isAuthenticated: false,
    userInfo: {},
    userId: null,
    errorMessage: null,
  };

  const REQUEST_PROPS_0 = {
    requests: null,
    getRequestsStarted: true,
    errorMessage: null,
    getRequestsAction: () => {},
  };

  const REQUEST_PROPS_1 = {
    requests: REQUESTS,
    getRequestsStarted: false,
    errorMessage: null,
    getRequestsAction: () => {},
  };

  const REQUEST_PROPS_2 = {
    requests: [],
    getRequestsStarted: false,
    errorMessage: null,
    getRequestsAction: () => {},
  };

  it.skip('blocks when not authorized', () => {
    const { getByTestId } = render(
      <Router>
        <Home {...{ ...AUTH_PROPS_3, ...REQUEST_PROPS_0 }} />
      </Router>
    );

    expect(getByTestId(TEST_IDS.AUTH.MODAL)).not.toBeNull();
  });

  it('has link to create a new request for reviewer', () => {
    const { getByTestId, container } = render(
      <Router>
        <Home {...{ ...AUTH_PROPS_0, ...REQUEST_PROPS_1 }} />
      </Router>
    );

    expect(getByTestId(TEST_IDS.APP.NEW_REQUEST)).toHaveTextContent('New Request');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('has link to create a new request for Requester when no open record exists', () => {
    const { getByTestId, container } = render(
      <Router>
        <Home {...{ ...AUTH_PROPS_1, ...REQUEST_PROPS_2 }} />
      </Router>
    );

    expect(getByTestId(TEST_IDS.APP.NEW_REQUEST)).toHaveTextContent('New Request');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('hide the link to create a new request for Requester when open record exists', () => {
    const { getByTestId, container } = render(
      <Router>
        <Home {...{ ...AUTH_PROPS_1, ...REQUEST_PROPS_1 }} />
      </Router>
    );

    expect(getByTestId(TEST_IDS.APP.HIDDEN_NEW_REQUEST)).toHaveTextContent(
      'You cannot create more requests until the current request is close.'
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  // TODO: create test for RequestList
  it('displays in different request list cases', () => {
    // loading records:
    const { getByTestId, rerender } = render(
      <Router>
        <Home {...{ ...AUTH_PROPS_0, ...REQUEST_PROPS_0 }} />
      </Router>
    );

    expect(getByTestId(TEST_IDS.APP.LOADER)).not.toBeNull();

    // show list of records:
    rerender(
      <Router>
        <Home {...{ ...AUTH_PROPS_0, ...REQUEST_PROPS_1 }} />
      </Router>
    );

    expect(getByTestId(TEST_IDS.REQUEST.FORM_LIST)).not.toBeNull();

    // show message saying no records found:
    rerender(
      <Router>
        <Home {...{ ...AUTH_PROPS_0, ...REQUEST_PROPS_2 }} />
      </Router>
    );

    expect(getByTestId(TEST_IDS.APP.EMPTY)).toHaveTextContent(
      'You do not have any realm requests yet, start a new request.'
    );
  });
});
