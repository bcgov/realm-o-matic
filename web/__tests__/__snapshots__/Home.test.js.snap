// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component blocks when not authed 1`] = `
<div>
  <h2>
    Welcome!
  </h2>
  <a
    data-testid="new-request-button"
    href="/Request/new"
  >
    <button
      class="ui button"
      data-testid="startRequest"
    >
      Start Request
    </button>
  </a>
  <p />
  <div
    class="ui small active text centered inline loader css-micwn4-StyledLoader e1c48y920"
    data-testid="load-indicator"
  >
    Loading requests...
  </div>
</div>
`;

exports[`Home Component has link to create a new request 1`] = `
<div>
  <h2>
    Welcome!
  </h2>
  <a
    data-testid="new-request-button"
    href="/Request/new"
  >
    <button
      class="ui button"
      data-testid="startRequest"
    >
      Start Request
    </button>
  </a>
  <p />
  <div
    class="ui small active text centered inline loader css-micwn4-StyledLoader e1eih1xm0"
    data-testid="load-indicator"
  >
    Loading requests...
  </div>
</div>
`;
